### ISSUE TITLE
[file icons] Adding default fileicon support to language contributions

### ISSUE THREAD

**Issue by ghost** [microsoft/vscode#14662]
It would be nice if extensions could specify custom file icons for the
file types it supports. For visual consistency this could be implemented
to only be used as a fallback if the user's selected theme does not specify
an icon.

**Comment by paniq** [comment]
yep. been very sad today to realize i couldn't add a file icon for my custom language. freedesktop icon themes has a very sensible implementation of all this, so i expected it to work here as well, but no. sad sad sad.

**Comment by AlexShemeshWix** [comment]
+1

**Comment by arnonkehat** [comment]
what's the process for getting a favicon into the themes that are packaged with vscode?

**Comment by aeschli** [comment]
@arnonkehat New icons for the built-in seti theme should be contributed directly to https://github.com/jesseweed/seti-ui

**Comment by RobPurcellUK** [comment]
Another vote for this. We have a YAML-based config for an internally-developed system, which I'm looking to create a syntax helper extension for. It'd be really good if I could specify the icon as well.

**Comment by DanTup** [comment]
It doesn't seem like seti-ui is active anymore, it's been over 7 months since the last update and many PRs have not been responding to. VS Code has already had to patch one icon directly because the fix hasn't been merged.

Right now, it seems like there isn't a way to get any new icon into VS Code 😞

**Comment by realpack** [comment]
My tree of files without icons due to the fact that I have to use extensions for work. 😢

**Comment by NavnathKumbhar** [comment]
I cloned the vscode repository which I forked. To add a new icon for my language extension, It seems that I have to modify sett.woff font file. Perhaps, I have to add a new glyph for my custom icon.

Is it acceptable change in vscode? Or can I add a image file for my custom icon inside extensions\theme-seti\icons\ ?

Thank you in advance.

**Comment by aeschli** [comment]
Modifying the woff file isn't the right approach. If you want to contribute to the seti theme directly, please make a PR against https://github.com/jesseweed/seti-ui

The idea here would be to add a mechnism to allow extensions to contribute a icon theme that extends an existing theme, instead of defining its own icon theme.

It's probanly not a lot of work
- extend the [`iconThemes` contribution point](https://github.com/Microsoft/vscode/blob/master/src/vs/workbench/services/themes/common/fileIconThemeStore.ts): Add an `extends` property
- when reading the icon theme definitions, also merge in the extenders

**Comment by DanTup** [comment]
@aeschli that would be great - maybe it's also worth having a way to provide an icon that doesn't need to explicitly list a theme to extend? (for ex. we have a standard Dart icon we'd prefer to use for all unknown themes, but for seti-ui we had to make a single-color version to fit in).

You may also need to consider if two extensions both contribute an extension to a theme with overlapping icons.

If this isn't likely to happen anytime soon, would you consider taking a forked version of seti-ui that has some of the PRs merged? It kinda sucks that right now that we can't get icons added :-(

**Comment by NavnathKumbhar** [comment]
@aeschli , Thank you for your feedback. Could you please elaborate your below points:

- extend the iconThemes contribution point: Add an extends property
- when reading the icon theme definitions, also merge in the extenders

Thank you in advance.

**Comment by aeschli** [comment]
>maybe it's also worth having a way to provide an icon that doesn't need to explicitly list a theme to extend
yes, we can try that out too.

**Comment by aeschli** [comment]
`iconThemes` is an existing contribution point that is used to contribute icon themes. See https://code.visualstudio.com/api/extension-guides/icon-theme for more information.
We could allow one additional property `"extends": "id" ` to signal that this is not a new icon theme but an extension of an existing theme with the id `id`.

**Comment by michaelblyons** [comment]
> We could allow one additional property `"extends": "id" ` to signal that this is not a new icon theme but an extension of an existing theme with the id `id`.

This would be great. Even better if `id` could

1. Be a list of IDs, allowing a single icon to add to multiple icon themes.
2. Allow a "special" object instead of an ID to assign that icon to _any_ addition icon theme that does not have an assignment for that file type.

**Comment by sentry07** [comment]
As a language extension developer, I don't want to extend any existing icon themes, I just want to provide icons for the files associated with my language. So instead of `contribues.iconThemes` in our package.json, if we could just use `contributes.icons` using the tags available in the icon-theme.json, that would be preferable. Like:

![image](https://user-images.githubusercontent.com/6489018/55184426-54e3eb00-514f-11e9-914e-bd2afaa2f0ec.png)

**Comment by michaelblyons** [comment]
@sentry07 That's what I would hope to have as a fall-back as I described in [No. 2 here][1]. And that's what I expect most people would use, but you could also contribute variants (No. 1) that tied in to specific, named icon themes as an alternative (Say there's a Seti-Monochrome, for instance).

[1]: https://github.com/Microsoft/vscode/issues/14662#issuecomment-473921069

**Comment by Hezkore** [comment]
Why was this closed?
I want to add a single icon for the little unknown language I've made an extension for.

**Comment by Hezkore** [comment]
Any news on this?

**Comment by SSJ9899TDylan** [comment]
Nope!

**Comment by joelday** [comment]
@aeschli Would you accept a PR for this based on @sentry07's suggested spec?

**Comment by aeschli** [comment]
Yes, the suggestion sounds good, I'd be happy to accept a PR. Maybe call it `default icons`?
Such icons would only show up if
- the current theme has no icon for the given file/language
- the theme shows icons
- shows specific icons (e.g. not in the minimal theme that only shows file/folder icons)

**Comment by Hezkore** [comment]
Any news on this @joelday?

**Comment by LadyCailin** [comment]
I actually started looking into it the other day. I didn’t get that far, cause it’s fairly decoupled. Maybe @aeschli can point us in the right direction. I see that the icons themselves are actually stored as a Unicode character in the private range. So the question is, how does one insert these icons into the.. font renderer? Font file? I’m not even sure what question to ask really, but somehow the icon is put somewhere so the renderer can turn a Unicode code point into the correct “character”. Anyways, I guess the configuration part, where the icon is loaded from the extension and placed “somewhere”, is fairly straightforward, but this part is less straightforward it seems.

**Comment by aeschli** [comment]
Yes whats needed is the same as there's in a file icon theme: https://code.visualstudio.com/api/extension-guides/icon-theme#icon-set-file
- A icon definition: declaring a new icon by name, either svg or font
- Font definitions, if fonts are used
- file/folder assocations
So essentially it's 3 new contribution points

We have a related discussion in https://github.com/microsoft/vscode/issues/84695#issuecomment-559232230.
Also there we're discussing ways for extensions to contribute named icons (to be used in markdowns, views, menus)

**Comment by pagnkelly** [comment]
so Any news about this?

**Comment by ImUrX** [comment]
@aeschli I was making an icon and discovered this doesnt exist, are you guys looking in doing a PR yourselves or can I contribute in the situation

**Comment by Antyos** [comment]
While I can't speak for everyone who's been watching this thread, it seems like we've all been waiting for _someone_ to finally make a pull request.

I've been working on an extension and similarly want to add an icon for OpenSCAD files. I've only just recently begun to learn TypeScript, so I'm not sure how much I can contribute. I would like to help bring this feature to VSCode, so I'll try to help where I can.

**Comment by supercruel** [comment]
Any news? Is this in development?

**Comment by SPGoding** [comment]
![extend](https://user-images.githubusercontent.com/15277496/99450323-40390580-28e6-11eb-8825-ace49f4751b1.gif)

Tried to implement this as [an idea of my extension](https://github.com/SPGoding/datapack-language-server/issues/341) has been blocked by this issue for months... I added an icon for `mcfunction` files in the test. Seems like it's working :eyes:

Will send a Pull Request soon.

**Comment by aeschli** [comment]
With the newly add icon contribution point (proposed API), we now can extend language definitions to name a default icon for the language. 
That icon would be used if the current file icon theme uses file icons and has not defined a icon for the language (or file extension).
```
    "languages": [
      {
        "id": "foo",
        "aliases": [ "FooLang" ],
        "extensions": [
          ".foo"
        ],
        "icon": "foo-icon"
        "configuration": "./language-configuration.json"
      }
    ]
 ```
 
`foo-icon` is an icon contributed like here: https://github.com/microsoft/vscode/issues/114942#issuecomment-773457375
 
Any takers? Otherwise I can give it a try in the coming weeks.

**Comment by Mai-Lapyst** [comment]
Hi,

im currently working on a patch, but I found out that there are actually several ways one could implement the feature,
so I am asking here for feedback which way would be the best way to implement it.

1. Since all file icons relevant css is generated in FileIconThemeData, one could simply generate fallback css if the icontheme don't provide icons for the language

2. another place to implement would be WorkbenchThemeService, since it's manages all themeing relevant data

3. way would be to change the behaviour of getIconClasses.ts so its also adds 'codicon-*' classes when its detect a modeId

2 & 3 could potentially involve updating treeView.ts Aligner.hasIcon() since it's only checks fileIconTheme.hasFileIcons and node.icon/node.iconDark, but it's unsure at the moment.

1 & 2 would duplicate some css, since the relevant code would be in a 'codicon-*' and a '{modeId}-lang-file-icon' class.

**Comment by itgenie98** [comment]
I've done some testing and got some updates on the 3 ways I explained above:

3. is out; since the codicon css has a lower precedence than the default icon from the icontheme, adding the 'codicon-*' class to the element dosn't do anything to the icon

2. can work; but the new rule overwrites the 'font-family' css attribute of the iconTheme, which will lead to incorrectly rendered icons since the icontheme's icon is not present in the font of the default icon. we can prevent this through setting the 'font-family' attribute on all css classes generated by a icontheme.

1. here we need to change a few function headers:
- _processIconThemeDocument
- FileIconThemeData.load
- FileIconThemeData.reload
so we can move the IModeService from WorkbenchThemeService to the processing function that generates the css.
this approach also woudn't need the patches as in 2. because we can test the theme if it has any overwrite for the any of the registered languages.

**Comment by Mai-Lapyst** [comment]
I've done some testing and got some updates on the 3 ways I explained above:

option 3: is out; since the codicon css has a lower precedence than the default icon from the icontheme, adding the 'codicon-*' class to the element dosn't do anything.

option 2: can work; but the new rule overwrites the 'font-family' css attribute of the iconTheme, which will lead to incorrectly rendered icons since the icontheme's icon is not present in the font of the default icon. we can prevent this through setting the 'font-family' attribute on all css classes generated by a icontheme.

option 1: here we need to change a few function headers:
- _processIconThemeDocument
- FileIconThemeData.load
- FileIconThemeData.reload
so we can move the IModeService from WorkbenchThemeService to the processing function that generates the css.
this approach also woudn't need the patches as in 2. because we can test the theme if it has any overwrite for the any of the registered languages.

**Comment by aeschli** [comment]
Option 1 sounds good to me. ok tot pass the IModeService to `load` and `reload`. You can also move `load` and `reload`to a new class FileIconThemeLoader (or a better name) that gets the services in its constructor.

**Comment by NavnathKumbhar** [comment]
Hello,
Has this issue been fixed already?
I tried below way as mentioned by @aeschli , but it is not working for me. 

>    "languages": [
      {
        "id": "foo",
        "aliases": [ "FooLang" ],
        "extensions": [
          ".foo"
        ],
        "icon": "foo-icon"
        "configuration": "./language-configuration.json"
      }
    ]

As this issue is in closed state, should I expect the fix in coming VSCode update?

**Comment by ImUrX** [comment]
@NavnathKumbhar you can see it has the insiders-released tag right now, so its only available on the insiders build. You can see that another issue was made too which is #140047

**Comment by aeschli** [comment]
The API is now final. See #140047 for the details.
